Add `elementReady.subscribe()`

qti3e opened the pull request:
This approach uses a `WeakMap` to store already seen elements (to prevent memory leak) in order to detect new elements, also because checking all the elements to find out new ones might be a bit heavy we do this in half of the `requestAnimationFrame`s, feedback is welcome.

Fixes #22


<!-- Issuehunt content -->

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Referenced issues

This pull request has been submitted to:
- [#22: Subscribe to new elements on the page](https://issuehunt.io/repos/62587127/issues/22)
---

IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>
<!-- /Issuehunt content-->

bfred-it commented on index.js:
```diff
@@ -54,4 +54,52 @@ const elementReady = (selector, {
+		checkNum += 1;
+	})();
+
+	return stop;
```
Maybe 
```suggestion
	return {stop};
```

So it matches the return type of `elementReady()`

bfred-it commented on index.js:
```diff
@@ -54,4 +54,52 @@ const elementReady = (selector, {
+		}
+
+		rafId = requestAnimationFrame(check);
+		checkNum += 1;
```
If this is confirmed to skip just one frame, it would become:

```suggestion
		checkNext = !checkNext;
```

bfred-it commented on index.d.ts:
```diff
@@ -24,6 +24,15 @@ declare namespace elementReady {
+    stop(): void;
+  };
+
 	type StoppablePromise<T> = Promise<T> & {
```
Perhaps

```ts
 	type StoppablePromise<T> = Promise<T> & Stoppable;
```

qti3e replied:
I actually wanted to do so, but I didn't want to change the doc comments.

qti3e commented:
@bfred-it Hey! I've made a small change and I guess this PR is ready to be merged.
Please let me know your opinion.

sindresorhus commented on index.js:
```diff
@@ -54,4 +54,52 @@ const elementReady = (selector, {
+		if (checkFrame) {
+			const allElements = target.querySelectorAll(selector);
+
+			for (let i = 0; i < allElements.length; ++i) {
```
Use a `for-of` loop.

sindresorhus commented on index.js:
```diff
@@ -54,4 +54,52 @@ const elementReady = (selector, {
 	return Object.assign(promise, {stop});
 };
 
+elementReady.subscribe = (selector, cb, {
```
Don't use abbreviations. `cb` => `callback`.

sindresorhus commented on index.test-d.ts:
```diff
@@ -12,4 +12,17 @@ expectType<elementReady.StoppablePromise<Element | undefined>>(promise);
 
+
+const {stop} = elementReady.subscribe(".unicorn", e => null);
+elementReady.subscribe(".unicorn", e => null, {target: document});
```
Use single-quotes

bfred-it replied:
Why doesn’t xo catch this?

sindresorhus replied:
XO doesn't run on TS files unless you enable it explicitly. https://github.com/xojs/xo/issues/373 will enable us to automatically lint `.d.ts` files too.

bfred-it replied:
Indeed, I meant to say "let's just enable XO for TS files on this repo"

sindresorhus replied:
Nah, I try to keep my boilerplate config as identical as possible between repos for ease of maintenance. I'd rather wait for XO.

sindresorhus commented on readme.md:
```diff
@@ -69,6 +69,36 @@ Stop checking for the element to be ready. The stop is synchronous and the origi
+
+Same as `elementReady` options.
+
+#### elementReady.subscribe()#stop()
```
?

sindresorhus commented on test.js:
```diff
@@ -187,3 +187,56 @@ test('ensure different promises are returned on second call with the same select
 });
+
+test('subscribe: check if elements are detected', async t => {
+	const e = () => {
```
Don't use one character variable names.

sindresorhus commented on readme.md:
```diff
@@ -69,6 +69,36 @@ Stop checking for the element to be ready. The stop is synchronous and the origi
+
+Detect elements as they are added to the DOM.
+
+Returns a stoppable to cancel subscription.
```
Not clear what a "stoppable" is.

sindresorhus commented on readme.md:
```diff
@@ -69,6 +69,36 @@ Stop checking for the element to be ready. The stop is synchronous and the origi
+### elementReady.subscribe(selector, cb, options?)
+
+Detect elements as they are added to the DOM.
+
```
Include some use-cases for this method.

sindresorhus commented on index.d.ts:
```diff
@@ -32,6 +41,44 @@ declare namespace elementReady {
+  /**
+  Called on each new element that matches the selector in the DOM.
+  */
+  type SubscribeCallback<T> = (el: T) => any;
```
This type is incorrect. According to the JS code, the return value is not used.
```suggestion
  type SubscribeCallback<T> = (element: T) => any;
```

sindresorhus commented on index.d.ts:
```diff
@@ -32,6 +41,44 @@ declare namespace elementReady {
+  Detect elements as they are added to the DOM.
+
+  @param selector - [CSS selector.](https://developer.mozilla.org/en-US/docs/Web/Guide/CSS/Getting_Started/Selectors)
+  @param cb Callback that is called for each new element.
```
```suggestion
  @param cb - Callback that is called for each new element.
```

sindresorhus commented on index.d.ts:
```diff
@@ -24,6 +24,15 @@ declare namespace elementReady {
+		/**
+		 Stop checking for new elements.
+
+		 Calling it multiple times does nothing.
```
Incorrect indentation.

sindresorhus commented on index.js:
```diff
@@ -54,4 +54,52 @@ const elementReady = (selector, {
+	stopOnDomReady = true,
+	timeout = Infinity
+} = {}) => {
+	const seen = new WeakSet();
```
```suggestion
	const seenElements = new WeakSet();
```

sindresorhus commented:
The implementation looks good, but I think you need to put more effort into the docs and TS definition.

qti3e commented:
@sindresorhus Sorry for all of this style-guide related issues, I wasn't quite familiar with your coding guide-lines,  anyway I've addressed all of your comments.

Also, I made a change to `SubscribeOptions`, I think  `SubscribeOptions.stopOnDomReady`'s default is better to be `false` as when a user calls this function they want to subscribe to all of the new elements, also they're probably going to use this function after the DOM ready event most of the time.

bfred-it commented:
1.  My usage would be exactly to catch item as the page loads, until the DOM ready
2. It’s best to match the options for the regular `elementReady`

qti3e commented:
ping @bfred-it

sindresorhus commented on index.test-d.ts:
```diff
@@ -12,4 +12,17 @@ expectType<elementReady.StoppablePromise<Element | undefined>>(promise);
 expectType<elementReady.StoppablePromise<SVGElement | undefined>>(elementReady('text'));
 
+
+const {stop} = elementReady.subscribe('.unicorn', e => null);
```
https://github.com/sindresorhus/element-ready/pull/24#discussion_r300868174

sindresorhus commented on index.test-d.ts:
```diff
@@ -12,4 +12,17 @@ expectType<elementReady.StoppablePromise<Element | undefined>>(promise);
+
+elementReady.subscribe('.unicorn', e => expectType<Element>(e));
+elementReady.subscribe('div', e => expectType<HTMLDivElement>(e));
+elementReady.subscribe('text', e => expectType<SVGElement>(e));
```
And these callbacks should not be inline arrow functions. They don't return anything so it's clearer with normal arrow functions.

fregante replied:
To be clear, Sindre means this by "normal arrow functions":

```js
elementReady.subscribe('text', e => {
	expectType<SVGElement>(e);
});
```

sindresorhus commented:
There are still some unresolved PR comments and some space indentation that needs to be converted to tab indentation.

sindresorhus commented:
@qti3e Ping back :)

qti3e commented:
@sindresorhus I've been busy working on something, sorry I forgot about this pr since it's been a long time.

sindresorhus commented on index.d.ts:
```diff
@@ -32,6 +41,44 @@ declare namespace elementReady {
+    selector: string,
+    callback: elementReady.SubscribeCallback<ElementName>,
+    options?: elementReady.Options
+  ): Stoppable;
```
Use tab-indentation.

sindresorhus commented on index.test-d.ts:
```diff
@@ -12,4 +12,23 @@ expectType<elementReady.StoppablePromise<Element | undefined>>(promise);
+	expectType<SVGElement>(element)
+});
+
+expectType<() => any>(stop);
```
This is not correct. It returns `void`.

sindresorhus commented on index.js:
```diff
@@ -49,4 +49,45 @@ const elementReady = (selector, {
+					continue;
+				}
+
+				callback(element);
```
Can you pass `undefined` instead of `null` (which `querySelectorAll` returns). To be consistent with our other usage of `undefined`.

sindresorhus commented on index.d.ts:
```diff
@@ -32,6 +41,44 @@ declare namespace elementReady {
+  /**
+  Called on each new element that matches the selector in the DOM.
+  */
+  type SubscribeCallback<T> = (element: T) => void;
```
The element is not guaranteed to exist so should be optional.

sindresorhus commented on test.js:
```diff
@@ -185,3 +185,56 @@ test('ensure different promises are returned on second call with the same select
 	t.is(prependElement(), await elementReady('.unicorn'));
 });
+
+test('subscribe: check if elements are detected', async t => {
```
There needs to be tests for the `timeout` and `stopOnDomReady` options with `subscribe` too.

sindresorhus commented on index.js:
```diff
@@ -49,4 +49,45 @@ const elementReady = (selector, {
+	const seenElements = new WeakSet();
+
+	let rafId;
+	let checkFrame = true;
```
This should should have a `should`, `is`, or `has` prefix. Not clear to me which one fits the best.

sindresorhus commented:
@qti3e Bump :)

sindresorhus commented:
Closing for lack of activity.
